### PR TITLE
Edit endpoint

### PR DESCRIPTION
**Pull request by Derinsola16**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,14 +4,22 @@
   "description": "An online quiz test",
   "main": "index.js",
   "dependencies": {
+    "bcrypt": "^3.0.6",
     "body-parser": "^1.19.0",
+    "dotenv": "^8.1.0",
```
Comment (holuborhee) [inline package.json]: what are you using `dotenv` for?

File: package.json [hunk 2]
```diff
@@ -4,14 +4,22 @@
...
     "express": "^4.17.1",
+    "express-session": "^1.16.2",
```
Comment (holuborhee) [inline package.json]: What's the express session for?

File: package.json [hunk 3]
```diff
@@ -4,14 +4,22 @@
...
+    "jsonwebtoken": "^8.5.1",
+    "passport": "^0.4.0",
```
Comment (holuborhee) [inline package.json]: Passport 🤷‍♂

File: package.json [hunk 4]
```diff
@@ -4,14 +4,22 @@
...
+    "passport-auth0": "^1.2.1",
```
Comment (holuborhee) [inline package.json]: this too?

File: server/controllers/login.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        })
+
+        if (users && users.length > 0)
+            allowLogin = bcrypt.compareSync(password, users[0].password)
+        
+        if(!allowLogin)
+            return res.status(401).send({message: 'Email and Password does not match!'});
+
+        const expiresIn = 24 * 60 * 60;
+             const accessToken = jwt.sign({ id: users.id }, SECRET_KEY, {
```
Comment (holuborhee) [inline server/controllers/login.js]: What type is `users`, is it an object?

File: server/controllers/login.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+                 expiresIn: expiresIn
```
Comment (holuborhee) [inline server/controllers/login.js]: Already mentioned that you don't need to repeat an object key to value when they are the same

File: server/controllers/question.js [hunk 7]
```diff
@@ -1,18 +1,63 @@
...
+        // match right option to question as answer
+        await Answer.create({ optionId: createdOptions[answer].id, questionId: question.id})
+
+        return res.status(201).send({ message: 'Question succesfully created'})
+    }
+
+
+static async list(req, res) { 
+    
+    const courses = await Course.findAll()
```
Comment (holuborhee) [inline server/controllers/question.js]: 🙄

File: server/controllers/question.js [hunk 8]
```diff
@@ -1,18 +1,63 @@
...
+
+    let course = courses
```
Comment (holuborhee) [inline server/controllers/question.js]: 👀

File: server/controllers/question.js [hunk 9]
```diff
@@ -1,18 +1,63 @@
...
+        const allQuestions = await Question.findAll({
+            order: [
+                [Sequelize.literal('RANDOM()')]
+            ],
+
+            limit: 5,
+
+            where: { courseId: course[5].id}
+        })
+   course++
```
Comment (holuborhee) [inline server/controllers/question.js]: 🙄

**Review by holuborhee — CHANGES_REQUESTED** [review]
What're all these 🤷‍♂ ?

File: server/controllers/quiz.js [hunk 10]
```diff
@@ -1,18 +1,55 @@
...
+           score += (answer.length * 20)
+        }
+        // crreate quiz
+        await Quiz.create({courseId, userId, score})
+
+        // return message and score as response
+        return res.status(201).send({ message: 'Quiz succesfully recorded', score })
+        
+        
+        //pair each question to to an option
```
Comment (holuborhee) [inline server/controllers/quiz.js]: remove all these commented lines of codes, line 31 - 40

File: server/controllers/user.js [hunk 11]
```diff
@@ -1,14 +1,31 @@
...
+    const password  =  bcrypt.hashSync(req.body.password, saltRounds);
     const role = 'CANDIDATE'
-    return User.create({email, password, role })
-    .then(userData => res.status(201)
-    .send({ success: true, message: 'User successfully created', userData })
-    ) 
-} 
-} 
+    //await User.create({email, password, role })
+        await User.create({role, email, password}, (err) => {
```
Comment (holuborhee) [inline server/controllers/user.js]: Please, can you explain all that I am seeing here. Line 14 - 26?

File: server/models/user.js [hunk 12]
```diff
@@ -1,12 +1,17 @@
 'use strict';
+const bcrypt = require('bcrypt');
```
Comment (holuborhee) [inline server/models/user.js]: does not look like you are using the bcrypt and salt here at all, remove them.

File: server/models/user.js [hunk 13]
```diff
@@ -1,12 +1,17 @@
...
 module.exports = (sequelize, DataTypes) => {
   const User = sequelize.define('User', {
     email: { type: DataTypes.STRING, allowNull: { args: false, msg: 'Please enter your email address' }, unique: { args: true, msg: 'Email already exists' }, },
     role: { type: DataTypes.ENUM('EXAMINER','CANDIDATE'), allowNull: { args: false, msg: 'Role cannot be empty'}, },
     password: { type: DataTypes.STRING, allowNull: {args: false, msg: 'Please enter a password' }, }
   }, {});
+  
   User.associate = function(models) {
     // associations can be defined here
+    // User.hasMany(models.Quiz)
```
Comment (holuborhee) [inline server/models/user.js]: remove this commented code

File: server/controllers/login.js (hunk 5 above)
Comment (Derinsola16) [reply server/controllers/login.js]: uers is the const users we created up

Comment (holuborhee) [reply server/controllers/login.js]: And what type is it? Object or Array?

File: server/controllers/login.js [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+        })
+
+        if (users && users.length > 0)
+            allowLogin = bcrypt.compareSync(password, users[0].password)
+        
+        if(!allowLogin)
+            return res.status(401).send({message: 'Email and Password does not match!'});
+
+        const expiresIn = 24 * 60 * 60;
+        delete users.password
```
Comment (holuborhee) [inline server/controllers/login.js]: Why are you still using users as though it is an object?

File: server/controllers/user.js [hunk 15]
```diff
@@ -1,14 +1,24 @@
...
+    const password  =  bcrypt.hashSync(req.body.password, saltRounds);
     const role = 'CANDIDATE'
-    return User.create({email, password, role })
-    .then(userData => res.status(201)
-    .send({ success: true, message: 'User successfully created', userData })
-    ) 
-} 
-} 
+
+      const users = await User.create({role, email, password})
```
Comment (holuborhee) [inline server/controllers/user.js]: Let your variable name reflect what it stands for, you are creating only one user why name it users?

File: server/controllers/user.js [hunk 16]
```diff
@@ -1,14 +1,24 @@
...
+
+        const expiresIn = 24 * 60 * 60;
```
Comment (holuborhee) [inline server/controllers/user.js]: Do not create any token for the user, this is not a login endpoint.

File: server/controllers/user.js [hunk 17]
```diff
@@ -1,14 +1,24 @@
...
+            const accessToken = jwt.sign({ id: users.id }, SECRET_KEY, {
+                     expiresIn
+                });
+
+        return res.status(201).send({ message: 'User successfully created', "user": users, "access_token": accessToken, "expires_in": expiresIn})
```
Comment (holuborhee) [inline server/controllers/user.js]: if you will send the user detail to the frontend delete the password from it

File: server/controllers/question.js [hunk 18]
```diff
@@ -1,18 +1,60 @@
...
+static async list(req, res) { 
+
+        const allQuestions = await Question.findAll({
+            order: [
+                [Sequelize.literal('RANDOM()')]
+            ],
+
+            limit: 5,
+
+            where: { courseId: 33}
```
Comment (holuborhee) [inline server/controllers/question.js]: you really need to explain what you are trying to do here

**Pull request merged**